src/scripts/workflow-collector.sh

```diff
@@ -44,58 +39,152 @@ else
+# Set FIRST_RUN to true to ensure all initial updates are sent.
+FIRST_RUN=true
+# While we still have jobs which are runnung.
+TIMEOUT=$(date -d "${PARAM_TIMEOUT_SECONDS} seconds")
+while true
 do
-    echo "looping: $i"
-    # fetch the job info
+  if [[ $(date) > $TIMEOUT ]];
+  then
+    echo "Monitoring loop exceeded timeout of $PARAM_TIMEOUT_SECONDS seconds. Breaking loop and ending the job."
+    break
+  fi
+
+  WF_DATA=$(curl -s "https://circleci.com/api/v2/workflow/$CIRCLE_WORKFLOW_ID/job?circle-token=${CIRCLE_TOKEN}")
+  WF_ITEMS=$(echo "$WF_DATA" | jq '.items')
+  WF_LENGTH=$(echo "$WF_ITEMS" | jq length)
+
+  WF_SL_PAYLOAD=$(curl -s "https://circleci.com/api/v2/workflow/$CIRCLE_WORKFLOW_ID?circle-token=${CIRCLE_TOKEN}" | jq '.')
+  WF_STATUS=$(echo "$WF_SL_PAYLOAD" | jq -r ".status")
+
+  if [[ "$WF_STATUS" != "running" ]];
+  then
+    echo "Workflow status no longer running. Now: ${WF_STATUS}. Breaking loop."
+    break
+  fi
+
+  # Check all jobs.
+  i="0"
+  while [ $i -lt "$WF_LENGTH" ]
+  do
     JOB_DATA=$(echo "$WF_ITEMS" | jq --arg i "$i" ".[$i]")
-    JOB_NUMBER=$(echo "$JOB_DATA" | jq ".job_number")
-    JOB_NAME=$(echo "$JOB_DATA" | jq ".name")
-    #####
-    # Send Job Data to SumoLogic
-    #####
-    if [ "$JOB_NUMBER" = "$CIRCLE_BUILD_NUM" ];
+    JOB_NUMBER=$(echo "$JOB_DATA" | jq -r ".job_number")
+    JOB_STATUS=$(echo "$JOB_DATA" | jq -r '.status')
+    JOB_NAME=$(echo "$JOB_DATA" | jq -r ".name")
+    if [[ "${JOB_NAME}" != "workflow-collector" ]];
     then
-        echo "This is the reporter job. Skipping"
-    else
-        echo "JOB: $JOB_NAME"
-        echo "JOB NUM: $JOB_NUMBER"
-        mkdir -p /tmp/sumologic-logs/
-        if [ "$JOB_NUMBER" = "null" ];
-        then
-            echo "Approval Job, skipping"
+      if ! [ "${job_status_array["${JOB_NAME}"]}" ];
+      then
+        echo "Job '$JOB_NAME' (job number: '$JOB_NUMBER') not tracked, adding to array with status of '$JOB_STATUS'."
+        job_status_array["${JOB_NAME}"]=$JOB_STATUS
+      fi
+
+      if [[ "$JOB_NUMBER" == "null" ]] && [[ "$JOB_STATUS" != "blocked" ]];
+      then
+        echo "'$JOB_NAME' has a job number of 'null' and status of '$JOB_STATUS'. What's gone wrong?"
+      elif [[ "$JOB_NUMBER" != "null" ]];
+      then
+        JOB_DATA_RAW=$(curl -s "https://circleci.com/api/v1.1/project/$VCS/$CIRCLE_PROJECT_USERNAME/$CIRCLE_PROJECT_REPONAME/$JOB_NUMBER?circle-token=${CIRCLE_TOKEN}")
+        JOB_STATUS=$(echo "$JOB_DATA_RAW" | jq -r '.status')
+        # Manually set job name as it is currently null
+        JOB_DATA_RAW=$(echo "$JOB_DATA_RAW" | jq --arg JOBNAME "$JOB_NAME" '.job_name = $JOBNAME')
+        JOB_STEP_NAMES=$(echo "$JOB_DATA_RAW" | jq '.steps' | jq .[] | jq '.name')
+        # Remove steps and circle_yml keys from object
+        JOB_DATA_RAW=$(echo "$JOB_DATA_RAW" | jq 'del(.circle_yml)' | jq 'del(.steps)')
+        JOB_COLLECTOR_NAME="Job Collector"
+        # Check if already monitored by a Job Collector.
```

Comment:
The first iteration `FIRST_RUN` is `true` so it should be sending the states as they come in. Subsequent iterations are just when changes are detected - is that what you are addressing?

Reply:
You are checking here whether a job's status has changed. If it has, then the curl works to send the changed data.
I was asking about what would have happened if the job's status didn't change? Are we still going to curl post that data? Because I don't see another if condition for that.

Reply:
The first state is sent due to the || FIRST RUN.